control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,268 @@
+func (r *GatewayClassController) isGatewayClassInUse(ctx context.Context, gc *gwv1beta1.GatewayClass) (bool, error) {
+	list := &gwv1beta1.GatewayList{}
+	if err := r.Client.List(ctx, list, &client.ListOptions{
+		FieldSelector: fields.OneTermEqualSelector(Gateway_GatewayClassIndex, gc.Name),
+	}); err != nil {
+		return false, err
+	}
+
+	return len(list.Items) != 0, nil
+}
+
+// validateParametersRef validates the ParametersRef field of the given GatewayClass
+// if it is set, ensuring that the referenced object is a GatewayClassConfig that exists.
+func (r *GatewayClassController) validateParametersRef(ctx context.Context, gc *gwv1beta1.GatewayClass, log logr.Logger) (didUpdate bool, err error) {
+	parametersRef := gc.Spec.ParametersRef
+	if parametersRef == nil {
+		return false, nil
+	}
+
+	if parametersRef.Kind != v1alpha1.GatewayClassConfigKind {
+		didUpdate, err := r.ensureCondition(ctx, gc, metav1.Condition{
+			Type:    invalidParameters,
+			Status:  metav1.ConditionTrue,
+			Reason:  configurationInvalid,
+			Message: fmt.Sprintf("Incorrect type for parametersRef. Expected GatewayClassConfig, got %q.", parametersRef.Kind),
+		})
+		if err != nil {
+			log.Error(err, "unable to update status")
+		}
+		return didUpdate, err
+	}
+
+	err = r.Client.Get(ctx, types.NamespacedName{Name: parametersRef.Name}, &v1alpha1.GatewayClassConfig{})
+	if k8serrors.IsNotFound(err) {
+		didUpdate, err := r.ensureCondition(ctx, gc, metav1.Condition{
+			Type:    invalidParameters,
+			Status:  metav1.ConditionTrue,
+			Reason:  configurationInvalid,
+			Message: fmt.Sprintf("GatewayClassConfig not found %q.", parametersRef.Name),
+		})
+		if err != nil {
+			log.Error(err, "unable to update status")
+		}
+		return didUpdate, err
+	}
+	if err != nil {
+		log.Error(err, "unable to fetch GatewayClassConfig")
+	}
+	return false, err
+}
+
+// ensureCondition ensures that the given condition is set on the GatewayClass.
+func (r *GatewayClassController) ensureCondition(ctx context.Context, gc *gwv1beta1.GatewayClass, condition metav1.Condition) (didUpdate bool, err error) {
+
+	// Update a condition the GatewayClass already has.
+	existsAt := -1
+	for i, c := range gc.Status.Conditions {
+		if condition.Type == c.Type {
+			existsAt = i
+		}
+	}
+	if 0 <= existsAt {
+		// Don't update if there is no change.
+		if equalConditions(condition, gc.Status.Conditions[existsAt]) {
+			return false, nil
+		}
+
+		// Update and set the time of the update.
+		gc.Status.Conditions[existsAt] = condition
+		gc.Status.Conditions[existsAt].LastTransitionTime = metav1.Now()
+
+		if err := r.Client.Status().Update(ctx, gc); err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}
+
+	// Add a new condition to the GatewayClass.
+	gc.Status.Conditions = append(gc.Status.Conditions, condition)
+	gc.Status.Conditions[len(gc.Status.Conditions)-1].LastTransitionTime = metav1.Now()
```

Comment:
so I think a number of things here can be simplified or might need some tweaks:

1. you should be able to just do the `LastTransitionTime` setting on the `condition` as it's passed in here, or move it to some sort of constructor for conditions
2. you'll want to set the `ObservedGeneration` on each of these conditions you're creating
3. I think you can just do the above scan looking for the `Type` value, if it exists you can do the `equalConditions` check and return early, if not, then do the condition setting and set a boolean that says `updated` and break out of the loop
4. If `!updated` then append the condition
5. do the update in one place at the end of the function.

If you want you could even remove the call to `Update` and do it in one go after both the parameter validation and the accepted condition setting so that we don't have to do another whole reconciliation.

Reply:
I addressed these a bit differently from how you recommended, but I wanted to keep the condition setting simple and was able to make it simpler.